# Optimizasyon

Şimdiye kadar az çok kod yazmayı öğrendik. Artık sorgulamamız gereken, yazdığımız kodun ne kadar etkili (efektif) olduğudur.

Aslında bilimsel programlama yaptığımız için, milisaniyelerin önemi var diyemeyiz. Bir işlemi $0.62$ saniyede yapmak ile $0.75$ saniyede yapmak arasında bir fark yok gibi.

Yoksa var mı? Söz konusu işlemi bir milyon defa yapmak zorunda kalmış olsak?

Hemen hesaplayalım:

İşlem süresi $0.62$ saniye iken:

In [5]:
islem_suresi = 0.62
tekrar_sayisi = 1_000_000

toplam_sure = islem_suresi * tekrar_sayisi

toplam_sure_saat = toplam_sure / 3600
toplam_sure_saat

172.22222222222223

İşlem süresi $0.62$ saniye iken:

In [6]:
islem_suresi = 0.75
tekrar_sayisi = 1_000_000

toplam_sure = islem_suresi * tekrar_sayisi

toplam_sure_saat = toplam_sure / 3600
toplam_sure_saat

208.33333333333334

:::{note}
Sayıyı ifade ederken ne yaptığımız gördünüz mü? ```_``` kullandım. Sayı ifade ederken ```_``` sayıyı etkilemez. Dolayısıyla büyük sayıları yazarken binler ayracı olarak kullanabiliriz.
:::

$0.62$ saniye süren işlemi bir milyon defa tekrarlayınca toplam süre yaklaşık $172$ saat iken, $0.75$ saniye süren işlemi bir milyon defa tekrarlayınca toplam süre yaklaşım $208$ saat oldu. Arada $1.5$ gün fark var.

```{image} ../images/shock_2.gif
:alt: Python resmi web sayfası
:class: bg-primary mb-1
:width: 400px
:align: center
```

Dolayısıyla bir işlemde $0.13$ saniyelik bir kazanımın ne denli önemli olabileceğini görüyoruz.

## Örnek

Bir milyon tane $0$'dan oluşan bir liste oluşturalım.

Bunun için iki farklı yöntem kullanalım ve süre farkına bakalım.

:::{note}
Süre farkını hesaplamak için ```time``` modülünü kullanalım.
:::

<hr>

Yöntem 1:

Bir döngü ile bir milyon adet $0$'ı listeye ekleyelim:

In [88]:
from time import time

baslangic = time()
sifirlar = []
for _ in range(1_000_000):
    sifirlar.append(0)
bitis = time()
print("Geçen süre = " + str(bitis - baslangic) + " Saniye")

Geçen süre = 0.1900186538696289 Saniye


:::{note}
Döngüde $[0, 999999]$ aralığındaki değerler üretildi. Fakat bu değerleri kullanmak istemiyorum. Sadece $1000000$ defa tekrar istiyorum. Dolayısıyla ```for```'un iterasyon değişkenine ```_``` diyorum.

```Python```'da ```_``` değişkeni, umursamadığımız değerler için kullanılır.
:::

Yöntem 2:

Şimdi ```list``` ve ```int``` tipi değişkenlerinin ```*``` özelliğini kullanalım:

In [89]:
from time import time

baslangic = time()
sifirlar = [0] * 1_000_000
bitis = time()
print("Geçen süre = " + str(bitis - baslangic) + " Saniye")

Geçen süre = 0.006022453308105469 Saniye


Gördünüz mü? Aradaki fark büyük. 

Yalnızca $0$'lardan oluşan dizinin oluşturma biçimini değiştirmekle, yaklaşık $21$ kat zaman kazandık.

Bunu daha gerçekçi bir örnek ile anlamaya çalışalım.

## Asal sayılar

Yalnızca 1 ve kendisine tam bölünebilen sayılardır.

<hr>

Örnek:

Verilen bir tam sayının asal olup olmadığını kontrol eden fonksiyonu yazınız.

Yöntem:
Verilen sayıyı (```n```), $[1, n]$ aralığındaki bütün sayılara bölelim. Tam bölme gerçekleştiği anda bir sayacın değerini arttıralım. Sonuçta sayacın değeri $2$'den büyük ise sayı asal değilmiş.

In [81]:
def asal(n: int) -> None:
    sayac = 0
    for i in range(1, n + 1):
        if n % i == 0:
            sayac += 1
            
    if sayac == 2:
        print(str(n) + " Asaldır")
    else:
        print(str(n) + " Asal değildir")
        
for i in range(2, 8):
    asal(i)

2 Asaldır
3 Asaldır
4 Asal değildir
5 Asaldır
6 Asal değildir
7 Asaldır


Şimdi çok büyük bir sayı için kodun ne kadar sürede yanıt verdiğine bakalım:

In [90]:
from time import time

def asal(n: int) -> None:
    sayac = 0
    for i in range(1, n + 1):
        if n % i == 0:
            sayac += 1
            
    if sayac == 2:
        print(str(n) + " Asaldır")
    else:
        print(str(n) + " Asal değildir")
        
baslangic = time()
asal(10000001)
bitis = time()
print("Geçen süre = " + str(bitis - baslangic) + " Saniye")

baslangic = time()
asal(10000019)
bitis = time()
print("Geçen süre = " + str(bitis - baslangic) + " Saniye")

10000001 Asal değildir
Geçen süre = 0.7879555225372314 Saniye
10000019 Asaldır
Geçen süre = 0.7771518230438232 Saniye


Şimdi bu fonksiyona optimizasyonlar yapalım.

### Çift sayılar

Çift sayıların tamamı $2$'ye tam bölündüğünden, $2$ dışındaki çift sayılardan kurtulalım:

In [91]:
def asal(n: int) -> None:
    if n == 2:
        print(str(n) + " Asaldır")
    else:
        if n % 2 == 0:
            print(str(n) + " Asal değildir")
        else:
            sayac = 0
            for i in range(1, n + 1, 2):
                if n % i == 0:
                    sayac += 1
            if sayac == 2:
                print(str(n) + " Asaldır")
            else:
                print(str(n) + " Asal değildir")

for i in range(2, 8):
    asal(i)

2 Asaldır
3 Asaldır
4 Asal değildir
5 Asaldır
6 Asal değildir
7 Asaldır


Şimdi bu fonksiyonun aldığı süreye bakalım

In [92]:
from time import time

def asal(n: int) -> None:
    if n == 2:
        print(str(n) + " Asaldır")
    else:
        if n % 2 == 0:
            print(str(n) + " Asal değildir")
        else:
            sayac = 0
            for i in range(1, n + 1, 2):
                if n % i == 0:
                    sayac += 1
            if sayac == 2:
                print(str(n) + " Asaldır")
            else:
                print(str(n) + " Asal değildir")


baslangic = time()
asal(10000001)
bitis = time()
print("Geçen süre = " + str(bitis - baslangic) + " Saniye")

baslangic = time()
asal(10000019)
bitis = time()
print("Geçen süre = " + str(bitis - baslangic) + " Saniye")

10000001 Asal değildir
Geçen süre = 0.4238407611846924 Saniye
10000019 Asaldır
Geçen süre = 0.369396448135376 Saniye


Fena değil. 

### Sayaç
Biz sayacı işlemlerin en sonunda kontrol ediyoruz. Aslında sayaç değeri $2$'yi aştığı anda daha fazla değere bakmamıza gerek kalmaz...

In [93]:
def asal(n: int) -> None:
    if n == 2:
        print(str(n) + " Asaldır")
    else:
        if n % 2 == 0:
            print(str(n) + " Asal değildir")
        else:
            sayac = 0
            for i in range(1, n + 1, 2):
                if n % i == 0:
                    sayac += 1
                    
                if sayac > 2:
                    break
            if sayac == 2:
                print(str(n) + " Asaldır")
            else:
                print(str(n) + " Asal değildir")
                
for i in range(2, 8):
    asal(i)

2 Asaldır
3 Asaldır
4 Asal değildir
5 Asaldır
6 Asal değildir
7 Asaldır


:::{note}
```break``` bir döngüyü durdurmak için kullanılır.
:::

Şimdi bu fonksiyonun aldığı süreye bakalım

In [94]:
from time import time

def asal(n: int) -> None:
    if n == 2:
        print(str(n) + " Asaldır")
    else:
        if n % 2 == 0:
            print(str(n) + " Asal değildir")
        else:
            sayac = 0
            for i in range(1, n + 1, 2):
                if n % i == 0:
                    sayac += 1
                    
                if sayac > 2:
                    break
            if sayac == 2:
                print(str(n) + " Asaldır")
            else:
                print(str(n) + " Asal değildir")

baslangic = time()
asal(10000001)
bitis = time()
print("Geçen süre = " + str(bitis - baslangic) + " Saniye")

baslangic = time()
asal(10000019)
bitis = time()
print("Geçen süre = " + str(bitis - baslangic) + " Saniye")

10000001 Asal değildir
Geçen süre = 0.039602041244506836 Saniye
10000019 Asaldır
Geçen süre = 0.5947058200836182 Saniye


:::{note}
Fonksiyonumuz asal olmayan sayılarda, asal sayılarda olduğundan çok daha hızlı çalışıyor.

Zira verilen sayı $2$ defa tam bölündüğünde döngü duruyor. Fakat sayı asal olunca döngünün sonuna kadar gitmek zorunda.
:::

Fena değil.

### Tüm sayılar

Olaya matematiksel açıdan bakalım.

Bir sayının bölenlerine bakalım.

```24``` için:

```1```, ```2```, ```3```, ```4```, ```6```, ```8```, ```12```, ```24```

Dikkat etmemiz gereken nokta:

Sayımız ```2```'ye tam bölünüyor. Bu aslında sayımızın ```12```'ye de tam bölündüğü anlamına gelir.

Aynı şekilde, sayımız ```3```'e tam bölünüyor. Bu da aslında sayımızın ```8```'e de tam bölündüğü anlamına gelir.

Kısacası ```24```'ün asal olup olmadığını anlamak için, sayımızın $[1, 24]$ aralığına bölünüp bölünmediğine değil de, $[1, \sqrt{24}]$'e bakmamız yeterli olur. Bu durumda sayımız herhangi bir değere bölünürse, ```sayı asal değildir``` diyebiliriz.

:::{note}
Sayının karekökünü alırken üste yuvarlayacağız. Garanti olsun diye.
:::

In [84]:
import math

def asal(n: int) -> None:
    if n == 2:
        print(str(n) + " Asaldır")
    else:
        if n % 2 == 0:
            print(str(n) + " Asal değildir")
        else:
            asaldir = True
            for i in range(3, math.ceil(math.sqrt(n)), 2):
                print(i)
                if n % i == 0:
                    asaldir = False
                    break
                    
            if asaldir:
                print(str(n) + " Asaldır")
            else:
                print(str(n) + " Asal değildir")
                
for i in range(2, 8):
    asal(i)

2 Asaldır
3 Asaldır
4 Asal değildir
5 Asaldır
6 Asal değildir
7 Asaldır


Şimdi bu fonksiyonun aldığı süreye bakalım

In [95]:
from time import time
import math

def asal(n: int) -> None:
    if n == 2:
        print(str(n) + " Asaldır")
    else:
        if n % 2 == 0:
            print(str(n) + " Asal değildir")
        else:
            asaldir = True
            for i in range(3, math.ceil(math.sqrt(n)), 2):
                if n % i == 0:
                    asaldir = False
                    break
                    
            if asaldir:
                print(str(n) + " Asaldır")
            else:
                print(str(n) + " Asal değildir")
                
baslangic = time()
asal(10000001)
bitis = time()
print("Geçen süre = " + str(bitis - baslangic) + " Saniye")

baslangic = time()
asal(10000019)
bitis = time()
print("Geçen süre = " + str(bitis - baslangic) + " Saniye")

10000001 Asal değildir
Geçen süre = 0.0 Saniye
10000019 Asaldır
Geçen süre = 0.0 Saniye


```{image} ../images/success.gif
:alt: Python resmi web sayfası
:class: bg-primary mb-1
:width: 400px
:align: center
```

Gördüğünüz gibi, bir az matematik, bir az mantık ile kodumuz daha uzun ve bir az daha karmaşık görünmesine rağmen çok daha hızlı çalışıyor.